controllers/spacerequest/spacerequest_controller_test.go

```diff
@@ -25,43 +27,733 @@ import (
+				spacetest.WithLabel(toolchainv1alpha1.ParentSpaceLabelKey, "jane"),
+				spacetest.WithCondition(spacetest.Ready()),
+				spacetest.WithSpecParentSpace("jane"),
+				spacetest.WithSpecTargetClusterRoles(srClusterRoles),
+				spacetest.WithStatusTargetCluster(member1.Name),
+				spacetest.WithTierName(sr.Spec.TierName))
+			hostClient := test.NewFakeClient(t, appstudioTier, parentSpace, subSpace)
+			ctrl := newReconciler(hostClient, member1)
+
+			// when
+			_, err := ctrl.Reconcile(context.TODO(), requestFor(sr))
+
+			// then
+			require.NoError(t, err)
+			// spacerequest exists with expected cluster roles and finalizer
+			spacerequesttest.AssertThatSpaceRequest(t, srNamespace.Name, sr.GetName(), member1.Client).
+				HasSpecTargetClusterRoles(srClusterRoles).
+				HasConditions(spacetest.Ready()).               // condition is reflected from space status
+				HasStatusTargetClusterURL(member1.APIEndpoint). // has new target cluster url
+				HasFinalizer()
+			// a subspace is created with the tierName and cluster roles from the spacerequest
+			spacetest.AssertThatSpace(t, test.HostOperatorNs, "jane-subs", hostClient).
+				HasSpecTargetClusterRoles(srClusterRoles).
+				HasConditions(spacetest.Ready()).
+				HasTier(sr.Spec.TierName).
+				HasParentSpace("jane") // the parent space is set as expected
+		})
+
+		t.Run("spacerequest created on member2", func(t *testing.T) {
+			// when
+			member1 := NewMemberClusterWithClient(test.NewFakeClient(t), "member-1", corev1.ConditionTrue)
+			member2 := NewMemberClusterWithClient(test.NewFakeClient(t, sr, srNamespace), "member-2", corev1.ConditionTrue) // space request and namespace are on member2
+			hostClient := test.NewFakeClient(t, appstudioTier, parentSpace)
+			ctrl := newReconciler(hostClient, member1, member2)
+			_, err := ctrl.Reconcile(context.TODO(), requestFor(sr))
+
+			// then
+			require.NoError(t, err)
+			// spacerequest exists with expected cluster roles and finalizer
+			spacerequesttest.AssertThatSpaceRequest(t, srNamespace.Name, sr.GetName(), member2.Client).
+				HasSpecTargetClusterRoles(srClusterRoles).
+				HasConditions(spacetest.Provisioning()).
+				HasFinalizer()
+			spacetest.AssertThatSpaces(t, hostClient).ContainsSubSpace(sr, parentSpace)
+		})
 	})
 
 	t.Run("failure", func(t *testing.T) {
-		// TODO implement error cases
+		// given
+		sr := spacerequesttest.NewSpaceRequest("jane",
+			srNamespace.GetName(),
+			spacerequesttest.WithTierName("appstudio"),
+			spacerequesttest.WithTargetClusterRoles(srClusterRoles))
+
+		t.Run("unable to find SpaceRequest", func(t *testing.T) {
+			member1 := NewMemberClusterWithClient(test.NewFakeClient(t), "member-1", corev1.ConditionTrue)
+			hostClient := test.NewFakeClient(t, appstudioTier)
+			ctrl := newReconciler(hostClient, member1)
+
+			// when
+			_, err := ctrl.Reconcile(context.TODO(), requestFor(nil))
```

Comment:
```suggestion
			_, err := ctrl.Reconcile(context.TODO(), requestFor(sr))
```

Reply:
this was testing the "unable to find SpaceRequest" case, probably we can drop this test.

Reply:
it's not found because the object was not added to the client, but the reconciliation should be triggered with some parameters, otherwise, the test doesn't make sense. It won't ever happen that the controller would get reconciled with empty parameters

Reply:
Good point, not adding it to the client is enough to enter the "unable to find SpaceRequest" path! Thanks!

Fixed in https://github.com/codeready-toolchain/host-operator/pull/770/commits/dc99b42393520b30615cc09069d1e759cd186950